ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/BlockSimulator.java

```diff
@@ -0,0 +1,423 @@
+ * class is responsible for simulating the execution of a block, which involves processing
+ * transactions and applying state overrides. It provides a way to test and validate the behavior of
+ * a block without actually executing it on the blockchain. The simulator takes into account various
+ * factors, such as the block header, transaction calls, and state overrides, to simulate the
+ * execution of the block. It returns a list of simulation results, which include the final block
+ * header, transaction receipts, and other relevant information.
+ */
+public class BlockSimulator {
+  private final TransactionSimulator transactionSimulator;
+  private final WorldStateArchive worldStateArchive;
+  private final ProtocolSchedule protocolSchedule;
+  private final MiningConfiguration miningConfiguration;
+
+  public BlockSimulator(
+      final WorldStateArchive worldStateArchive,
+      final ProtocolSchedule protocolSchedule,
+      final TransactionSimulator transactionSimulator,
+      final MiningConfiguration miningConfiguration) {
+    this.worldStateArchive = worldStateArchive;
+    this.protocolSchedule = protocolSchedule;
+    this.miningConfiguration = miningConfiguration;
+    this.transactionSimulator = transactionSimulator;
+  }
+
+  /**
+   * Processes a list of BlockStateCalls sequentially, collecting the results.
+   *
+   * @param header The block header for all simulations.
+   * @param blockStateCalls The list of BlockStateCalls to process.
+   * @return A list of BlockSimulationResult objects from processing each BlockStateCall.
+   */
+  public List<BlockSimulationResult> process(
+      final BlockHeader header, final List<? extends BlockStateCall> blockStateCalls) {
+    try (final MutableWorldState ws =
+        worldStateArchive
+            .getMutable(header, false)
+            .orElseThrow(
+                () ->
+                    new IllegalArgumentException(
+                        "Public world state not available for block " + header.toLogString()))) {
+      return process(header, blockStateCalls, ws);
+    } catch (IllegalArgumentException e) {
+      throw e;
+    } catch (final Exception e) {
+      throw new RuntimeException("Error simulating block", e);
+    }
+  }
+
+  /**
+   * Processes a list of BlockStateCalls sequentially, collecting the results.
+   *
+   * @param header The block header for all simulations.
+   * @param blockStateCalls The list of BlockStateCalls to process.
+   * @param worldState The initial MutableWorldState to start with.
+   * @return A list of BlockSimulationResult objects from processing each BlockStateCall.
+   */
+  public List<BlockSimulationResult> process(
+      final BlockHeader header,
+      final List<? extends BlockStateCall> blockStateCalls,
+      final MutableWorldState worldState) {
+    List<BlockSimulationResult> simulationResults = new ArrayList<>();
+    for (BlockStateCall blockStateCall : blockStateCalls) {
```

Comment:
When you are simulating multiple blocks, does it make sense that each of the blockStateCalls has it's own blockOverrides?